// @tanstack/pacer/dist/esm/index.js
import { AsyncBatcher, asyncBatch } from "./async-batcher.js";
import { AsyncDebouncer, asyncDebounce } from "./async-debouncer.js";
import { AsyncQueuer, asyncQueue } from "./async-queuer.js";
import { AsyncRateLimiter, asyncRateLimit } from "./async-rate-limiter.js";
import { AsyncThrottler, asyncThrottle } from "./async-throttler.js";
import { Batcher, batch } from "./batcher.js";
import { Debouncer, debounce } from "./debouncer.js";
import { Queuer, queue } from "./queuer.js";
import { RateLimiter, rateLimit } from "./rate-limiter.js";
import { Throttler, throttle } from "./throttler.js";
import { createKey, isFunction, parseFunctionOrValue } from "./utils.js";
import { pacerEventClient } from "./event-client.js";
export {
  AsyncBatcher,
  AsyncDebouncer,
  AsyncQueuer,
  AsyncRateLimiter,
  AsyncThrottler,
  Batcher,
  Debouncer,
  Queuer,
  RateLimiter,
  Throttler,
  asyncBatch,
  asyncDebounce,
  asyncQueue,
  asyncRateLimit,
  asyncThrottle,
  batch,
  createKey,
  debounce,
  isFunction,
  pacerEventClient,
  parseFunctionOrValue,
  queue,
  rateLimit,
  throttle
};
